# Project Description


Due to the global pandemic, more people were at home and reading books. This attracted the attention of startups that rushed to develop new apps for book lovers. The given database of one of the service competing in market consists data on books , publishers etc., This information must be used to generate a value proposition for a new product

### Data Description:

books:
1. Contains data on books:
2. book_id
3. author_id
4. title
5. num_pages — number of pages
6. publication_date
7. publisher_id

authors:
1. Contains data on authors:
2. author_id
3. author

publishers:
1. publisher_id
2. publisher

ratings:
1. rating_id
2. book_id
3. username — the name of the user who rated the book
4. rating

reviews:
1. review_id
2. book_id
3. username — the name of the user who reviewed the book
4. text — the text of the review



In [1]:
!pip install psycopg2-binary

Defaulting to user installation because normal site-packages is not writeable


In [2]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

In [3]:
db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
#reading books dataset

query = 'SELECT * FROM books'

books = pd.io.sql.read_sql(query, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
#reading authors dataset

query = 'SELECT * FROM authors'

authors = pd.io.sql.read_sql(query, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
#reading publishers


query = 'SELECT * FROM publishers'

publishers = pd.io.sql.read_sql(query, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [7]:
#reading ratings


query = 'SELECT * FROM ratings'

ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [8]:
#reading reviews


query = 'SELECT * FROM reviews'

reviews = pd.io.sql.read_sql(query, con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### Tasks

In [9]:
query_list = {'TASK 1 - Find the number of books released after January 1, 2000.' : "SELECT count(*) FROM books where publication_date > cast('2000-01-01' as date)",
             'TASK 2 - Find the number of user reviews and the average rating for each book.' : 'SELECT books.title, COUNT(DISTINCT review_id) as user_review, AVG(rating) as avg_rating from reviews INNER JOIN ratings on reviews.book_id = ratings.book_id INNER JOIN books ON books.book_id = reviews.book_id group by books.book_id ORDER BY user_review DESC ',
             'TASK 3 - Identify the publisher that has released the greatest number of books with more than 50 pages ': '''SELECT  publisher,count(books.book_id) 
                FROM publishers INNER JOIN books ON publishers.publisher_id = books.publisher_id 
                WHERE books.num_pages > 50
                GROUP BY publishers.publisher
                ORDER BY COUNT(books.book_id) DESC 
                LIMIT 1''',
              'TASK 4 - Identify the author with the highest average book rating (look only at books with at least 50 ratings)':
                '''SELECT author, AVG(rating)
                FROM books, authors, ratings
                WHERE books.book_id = ratings.book_id
                AND authors.author_id = books.author_id
                group by authors.author
                HAVING count(rating) > 50
                LIMIT 1''',
              'TASK 5 - Find the average number of text reviews among users who rated more than 50 books' : '''
            SELECT AVG(count)
            FROM(SELECT count(review_id)
            FROM reviews 
            WHERE username in 
            (SELECT username from ratings
            GROUP BY username
            HAVING count(rating_id)>50)
            GROUP BY username) a
            '''}


for key in query_list:
    print(key)
    print('')
    print(pd.io.sql.read_sql(query_list[key], con = engine))
    print('')
    print('')
    print('')

TASK 1 - Find the number of books released after January 1, 2000.

   count
0    819



TASK 2 - Find the number of user reviews and the average rating for each book.

                                   title  user_review  avg_rating
0                Twilight (Twilight  #1)            7    3.662500
1                               The Road            6    3.772727
2                         The Book Thief            6    4.264151
3                       The Glass Castle            6    4.206897
4                    Water for Elephants            6    3.977273
..                                   ...          ...         ...
989    Naked Empire (Sword of Truth  #8)            1    3.500000
990                  Moo  Baa  La La La!            1    3.000000
991  Merrick (The Vampire Chronicles #7)            1    4.000000
992                            Babyville            1    3.500000
993                           Lysistrata            1    4.000000

[994 rows x 3 columns]



TASK 3 - Iden

### Conclusion

- There were a total of 819 books released after jan 1st 2000
- The number of user reviews and the average rating for each book was found. Books should be considered in a way that both the user reviews as well as rating is more
- Penguin Books is the publisher that released the greatest number of books consisting of more than 50 pages
- There are a total of 43 authors who have equally the highest average rating for their books 
- The average number of text reviews among those users who rated more than 50 books is 24.33